include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Record a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value is the value to store for a Gauge or increment for a histogram or Counter.
+   */
+  virtual WasmResult recordMetric(uint32_t /* metric_id */, uint64_t /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Get the current value of a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value_ptr is a location to store the value of the metric.
+   */
+  virtual WasmResult getMetric(uint32_t /* metric_id */, uint64_t * /* value_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Properties
+
+  /**
+   * Get the value of a property.  Some properties are proxy-independent (e.g. ["plugin_root_id"])
+   * while others can be proxy-specific.
+   * @param path is a sequence of strings describing a path to a property.
+   * @param result is a location to write the value of the property.
+   */
+  virtual WasmResult getProperty(string_view /* path */, std::string * /* result */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Set the value of a property.
+   * @param path is a sequence of strings describing a path to a property.
+   * @param value the value to set.  For non-string, non-integral types, the value may be
+   * serialized..
+   */
+  virtual WasmResult setProperty(string_view /* key */, string_view /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Continue
+
+  /**
+   * Continue processing a request e.g. after returning ProxyAction::Pause.
+   */
+  virtual void continueRequest() { unimplemented(); }
+
+  /**
+   * Continue processing a response e.g. after returning ProxyAction::Pause.
+   */
+  virtual void continueResponse() { unimplemented(); }
+
+  /**
+   * Respond directly to an HTTP request.
+   * @param response_code is the response code to send.
+   * @param body is the body of the response.
+   * @param additional_headers are additional headers to send in the response.
+   * @param grpc_status is an optional gRPC status if the connection is a gRPC connection.
+   * @param details are details of any (gRPC) error.
+   */
+  virtual void sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                                 Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                                 string_view /* details */) {
+    unimplemented();
+  }
+
+  // Shared Data
+
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
```

Comment:
Why do we have complex low-level CAS semantics? Can't we just have the return value from the operation say "yes/no"?

Reply:
I am not sure what you are proposing.  This is the standard CAS system AFAIK.  Do you have an example of an alternative API you would prefer?

Reply:
I guess it would be good to add a section on concurrency model to the ABI spec, I don't recall seeing one.

Reply:
We should add one.  The model is that each VM is single threaded, so inter-VM communication is (in general) cross-thread.